Edited formatting in index_md

cilwerner opened the pull request:
Discussed index.md has been addressed, and some additions made to episodes_02 and 01

cilwerner committed:
spelling corrected

Commit:
Created table of LAMMPS features supported by KOKKOS from list and added small everyday example for explaining benchmarking and scaling.

Commit:
Merge branch 'gh-pages' of https://github.com/cilwerner/tuning_lammps into gh-pages
updated git page

Commit:
Added working cube picture

Commit:
Added Prithwish material episode 02 down to L150.

Commit:
Added input LAMMPS file

Commit:
Reduced cube size in episode_01 and testing new snippet abstractions

Commit:
Testing abstractions with config in snippets library directly

Commit:
Abstraction test 1

Commit:
edited yml file for testing abstractions

Commit:
Changed index.md to be less wordy. Original version stored in markdown file in files directory.

Commit:
Removed some material from index.md.

Commit:
index_orig updated in files

cilwerner committed:
Updated formatting

cilwerner committed:
Reverted to commit (9028c82) for ep02, removing content

Reasons for this were to keep pull request topics separate for easier review. Episode 2 has been reverted to skeleton code. See pull request from prithwish-ichec. Corrected any spelling mistakes.

ocaisa commented on _config.yml:
```diff
@@ -60,6 +60,36 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
+host_location: "JSC, Julich Supercomputing Centre"
+#host_homedir: "/ichec/home/course01"  # watch out due to course accounts
+host_prompt: "[user@laptop ~]$"
+sched_name: "SBATCH" # Check
```
I'd probably use SLURM here

cilwerner replied:
Changed to SLURM.

ocaisa commented on _config.yml:
```diff
@@ -60,6 +60,36 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
+sched_status: "sstat"
+sched_flag_user: "-u yourUsername"
+sched_del: "scancel"
+sched_interactive: "sbatch"
```
For interactive session in SLURM I think you usually need something like
```
srun --pty bash -i
```
see https://apps.fz-juelich.de/jsc/hps/jureca/quickintro.html#interactive-sessions

cilwerner replied:
wrote `srun --pty /bin/bash -i` based on link above

ocaisa commented on _config.yml:
```diff
@@ -60,6 +60,36 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
+sched_flag_name: "-N" # EDITME
+sched_hist: "" # EDITME
+sched_flag_histdetail: "" # EDITME 
+module_load: "module load LAMMPS/9Jan2020-cuda intel-para/2019a Stages/Devel-2019a"
```
I think for `module load` and `module usage` I would just use something like
```
execution_environment_kokkos: "module use /usr/local/software/jureca/OtherStages
module load Stages/Devel-2019a
module load intel-para/2019a LAMMPS/9Jan2020-kokkos"
execution_environment_gpu:"module use /usr/local/software/jureca/OtherStages
module load Stages/Devel-2019a
module load intel-para/2019a LAMMPS/9Jan2020-cuda"
```

ocaisa commented on _config.yml:
```diff
@@ -60,6 +60,36 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
+sched_flag_histdetail: "" # EDITME 
+module_load: "module load LAMMPS/9Jan2020-cuda intel-para/2019a Stages/Devel-2019a"
+module_usage: "module use /usr/local/software/jureca/OtherStages"
+run_openmp: "srun"
```
You don't strictly need `srun` here but it will set up the process pinning for you so is perhaps not a bad thing

cilwerner replied:
Will check with Prithwish on the correct command

ocaisa commented on _config.yml:
```diff
@@ -60,6 +60,36 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
+module_load: "module load LAMMPS/9Jan2020-cuda intel-para/2019a Stages/Devel-2019a"
+module_usage: "module use /usr/local/software/jureca/OtherStages"
+run_openmp: "srun"
+run_mpi: ""
```
You do need `srun` here though, but srun on its own is probably not enough. For mpirun you would also need `mpirun -n {number of cores}`, though I am not sure how we would handle this.

cilwerner replied:
if it is left as `mpirun -n` and then we can have a generic core count (eg 4)

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -20,16 +20,10 @@ keypoints:
 ## What is software performance?
 
-Before getting into the software side of things, lets take a few steps back. Performance is a generic term that applies to may factors of our lives, from physical to mental and engineering. However with software performance the emphasis isn't always on the most powerful machine, it is how you best utilise the power that you have. Say you are the chef in a restaurant and every dish that you do is  ...
+Before getting into the software side of things, lets take a few steps back. Performance is a generic term that applies to may factors of our lives, from physical to mental and engineering. However with software performance the emphasis isn't always on the most powerful machine, it is how you best utilise the power that you have. Say you are the chef in a restaurant and every dish that you do is  ...
```
```
Before getting into the software side of things, lets take a few steps back. Performance is a generic term that
applies to many factors of our lives, such as our own physical and mental performance, as well as a
work-oriented meaning when we consider fields such as engineering. With software performance the emphasis
isn't centred running on the most powerful machine, it is how you best utilise the power that you have. Say you
are the chef in a restaurant and every dish that you do is perfect. You would be able to produce a set 7 course
meal for a table of 3-6 without too much difficulty. If your restaurant is full though, it becomes more difficult, as
people would be waiting many hours for their food. However if you had a team of 6 additional chefs with you
with whom you can delegate tasks and who are assisting as well as communicating with each other, you have a
much higher chance of getting the food out on time and coping with a large workload. Most likely each chef will
end up primarily concentrating on doing one course.
```
I like the example.

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -20,16 +20,10 @@ keypoints:
-Before getting into the software side of things, lets take a few steps back. Performance is a generic term that applies to may factors of our lives, from physical to mental and engineering. However with software performance the emphasis isn't always on the most powerful machine, it is how you best utilise the power that you have. Say you are the chef in a restaurant and every dish that you do is  ...
+Before getting into the software side of things, lets take a few steps back. Performance is a generic term that applies to may factors of our lives, from physical to mental and engineering. However with software performance the emphasis isn't always on the most powerful machine, it is how you best utilise the power that you have. Say you are the chef in a restaurant and every dish that you do is  ...
 
 It's the same with software performance, if you have one process running doing one thing there is usually never an issue. But if you have one doing several things, it will obviously take much longer. Furthermore if you hade several workers working on several tasks, it will take a fraction of the time. That is the essence of software performance, using the resources you have to their best capabili ...
```
```
It's the same with software performance, if you have one process running doing one thing there is usually never an
issue. But if you have one doing several things, it will obviously take much longer. If you have access to several workers to whom you can assign tasks or effort, it will take a fraction of the time. That is the essence of software performance, using the resources you have to the best of their capabilities.
```

ocaisa reviewed (changes requested):
Not quite finished yet but I have to go until 1pm. Will take another look then (12pm your time)

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -38,7 +32,7 @@ Computing works in a similar way, thankfully not to that level of detail where o
 >
 > Imagine you had a 40 x 40 x 40 box like the one below, divided up into smaller boxes, each measuring 1 x 1 x 1. If you wanted and you wanted to simulate what was happening inside each smaller box for 10 hours each How long will the example below take to run on one core?
```
Remove `and you wanted`

ocaisa replied:
In the solution below you mention 10-12 hours, but you don't say why it could takes more than 10 hours. This is perhaps an opportunity to introduce the idea of *linear scaling* and why you typically won't achieve it due to overhead.

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -55,11 +49,11 @@ Computing works in a similar way, thankfully not to that level of detail where o
 
 There are a number of key terms in computing when it comes to understanding performance and expected duration of tasks. The most important of these are flops, walltime and cpuh
```
I'd change the order here: `flops, cpuh and walltime` to end with the one that they should care about the most as *application users*. As *application users* they are not going to be able to directly impact the flops.  The cpuh they can influence by options to trigger different code paths in the application, which *might* result in lower walltime *for the same resources* and *for their use case*. The other way to change walltime is to tweak the resources being used.

ocaisa commented on _episodes/02-benchmark-and-scaling.md:
```diff
@@ -37,11 +37,9 @@ keypoints:
 (FIXME)
 
-> ## Plotting performacne and number of cores
+> ## Plotting performanne and number of cores
```
`performance`

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -141,14 +136,14 @@ The `--time` variable used in the exercise is the amount of CPU hours requested,
-> > As you can see, there are a lot of right answers, however some methods work better than others, and it can entirely depend on the problem you are trying to solve.
-> >
+>> ## Solution
+>> 1. Yes, the more cores and nodes you have, the more work can be distributed across those cores.
```
There's a caveat here: *if your application is capable of using them*

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -141,14 +136,14 @@ The `--time` variable used in the exercise is the amount of CPU hours requested,
+>> ## Solution
+>> 1. Yes, the more cores and nodes you have, the more work can be distributed across those cores.
+>> 2. No, increasing simulation walltime and CPUh only increases the duration the code will run for. It does not improve the code performance.
+>> 3. Yes, making the code more efficient by cutting out unnecessary loops and bottlenecks can greatly increase performance.
```
I wouldn't give an example here, performance optimisation is a minefield, say `making the code more efficient by identifying and improving bottlenecks can greatly increase performance`

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -141,14 +136,14 @@ The `--time` variable used in the exercise is the amount of CPU hours requested,
+>> 1. Yes, the more cores and nodes you have, the more work can be distributed across those cores.
+>> 2. No, increasing simulation walltime and CPUh only increases the duration the code will run for. It does not improve the code performance.
+>> 3. Yes, making the code more efficient by cutting out unnecessary loops and bottlenecks can greatly increase performance.
+>> 4. Yes, the faster the computer, the faster the code can run, however, this may not always be possible logistically.
```
Historically this was what happened but modern HPC machines worry a lot about their electricity bill. Slower CPUs, but lots of them, is a more power-efficient way to achieve higher *overall* performance.

ocaisa replied:
Another caveat here is that sometimes it is not the speed of the CPU, but how fast you can get the data it needs to it. If your CPU spends time idle waiting for data to arrive (*I/O bound* calculations), a faster CPU does not help.

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -164,11 +159,12 @@ If these terms are unfamiliar now, don't worry as we will cover them over the du
+>> 1. Yes, MPI can enable you to split your code into multiple processes distributed over multiple cores, known as parallel programming.
+>> 2. Yes, like MPI this is also parallel programming, but deals with threads, by splitting a process into multiple threads, each thread using a single core.
+>> 3. Yes, that is their purpose. Different libraries run on different architectures however.
+>> 4. Yes, GPUs are better at handling multiple tasks, whereas a CPU is better running singular tasks quickly.
```
The truth in this statement is changing, even Kay has 40 cores per node.

ocaisa commented on _episodes/01-why-bother-performance.md:
```diff
@@ -164,11 +159,12 @@ If these terms are unfamiliar now, don't worry as we will cover them over the du
+>> 2. Yes, like MPI this is also parallel programming, but deals with threads, by splitting a process into multiple threads, each thread using a single core.
+>> 3. Yes, that is their purpose. Different libraries run on different architectures however.
+>> 4. Yes, GPUs are better at handling multiple tasks, whereas a CPU is better running singular tasks quickly.
+>> 5. No, if you have a simulation that needs to be run from start to completion, splitting the code into segments won't be of any benefit and will waste compute resources.
```
Well, I disagree a little bit here depending on what you mean by splitting into segments. Compilers are quite clever, making code more read-able through refactoring will most likely not waste resources because of this. What it will do is save *human resources* for someone who comes along and wants to modify the code (including the original developer). Human resources are far more valuable than CPU resources, so even if this came with a 10% performance price tag it is worth it,

ocaisa commented on _episodes/05-accelerating-lammps.md:
```diff
@@ -34,46 +34,20 @@ But to obtain the Kokkos accelartion, LAMMPS needs to be built using Kokkos for
 
 The list of LAMMPS features that is supported by Kokkos is given below:
```
If we can just provide a link here do so, this list is a moving target, we should use a link so our information is up to date.

ocaisa commented on _extras/snippets_library/_config_options.yml:
```diff
@@ -0,0 +1,27 @@
+# JSC, Julich Supercomputing Centre: Jureca + SBATCH
+host_name: "Jureca"
+host_login: "kay.ichec.ie"
```
`jureca.fz-juelich.de`

ocaisa commented on _extras/snippets_library/_config_options.yml:
```diff
@@ -0,0 +1,27 @@
+host_login: "kay.ichec.ie"
+host_location: "JSC, Julich Supercomputing Centre"
+host_homedir: "/ichec/home/course01"  # watch out due to course accounts
+host_prompt: "[user@laptop ~]$"
```
That's not the default prompt on JURECA

ocaisa commented on files/index_orig_02_04_20.md:
```diff
@@ -0,0 +1,60 @@
+## Why should I take this course?
+
+>Before you start the class, type a few sentences about why you are here.
```
Type? Where? Maybe talk to your neighbour instead.

ocaisa replied:
Ah you are archiving this file, ok, but we will need to clear this out later

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
->   * How slow is running LAMMPS on my HPC than its standard performance?
+* What is meant by the term 'performance' of a software?
+* What is the metric of measuring this performance?
+* How do I measure standard performance of a software?
```
`How do I measure the performance of a piece of software?`

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
-At this point, I assume that all of you have some level of experiences of running [LAMMPS](https://lammps.sandia.gov/) software on a HPC system. A HPC enable us to run jobs that would either be impossible or much slower either in a Desktop or a laptop. The phrase 'much slower' in the previous sentence is not much informative. The first question that arises here is:
->   * How slow is running LAMMPS on my HPC than its standard performance?
+* What is meant by the term 'performance' of a software?
+* What is the metric of measuring this performance?
```
delete `this`

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
+* What is meant by the term 'performance' of a software?
+* What is the metric of measuring this performance?
+* How do I measure standard performance of a software?
+* How slow is LAMMPS running on my HPC relative to its standard performance?
```
This questions is not clear, do you mean:
```
What performance should I expect when running LAMMPS on my HPC system and how do I achieve it?
```

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
+* What is the metric of measuring this performance?
+* How do I measure standard performance of a software?
+* How slow is LAMMPS running on my HPC relative to its standard performance?
+* **If a software performance is not optimal in my system, is there something that can I do to accelerate it?**
```
`optimal` is a dangerous word, there are too many factors at play, `acceptable` is a better option since it acknowledges that there is probably always something we could do a little better. There is a good concept that comes from the VI-HPS workshops, which (para-phrasing) is: 80% of the performance improvement comes with 20% of the effort, that last 20% may not be worth it, know when to stop.

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
 
-and finally,
->   * if a software performance is not optimal in my system, is there something that can I do to accelerate it?
+An HPC is a complex computing platform that has several hardware components and we often talk about its central processing units (CPU), primary and secondary memories and various interconnects. The processing units do the calculations, and there are many different types/generations. The most basic CPU has only one core. This means that it can do only one job at a time.
```
Also need to include accelerators

ocaisa replied:
Single CPU cores probably don't even exist anymore :P

ocaisa replied:
Probably it is best to pull a short paragraph from the HPC intro lesson here, see https://github.com/hpc-carpentry/hpc-intro/blob/gh-pages/_episodes/12-cluster.md#what-is-an-hpc-system

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
 # Workforces on a HPC
 HPC is a complex computing platform that has several hardware components and we often talk about its central processing units (cpu), primary and secondary memories and various interconnects. The processing units are the ones that actually do the calculations and if you study the evaluation of these processing units you would be quite amazed to see how many different types/generations they have. T ...
 
 In recent days, we talk about multi-core cpus and graphical processing units (gpu). Naturally, the muti-core cpus have many processing units that can work in parallel. Similarly the gpus have many processing units as well but they work differenty than a cpu. If you look at the infrastructure of a modern HPC, you would find that it is a heterogeneous computing system since its computing workforce  ...
```
Duplicated from above

ocaisa commented on index.md:
```diff
@@ -4,61 +4,38 @@ root: .  # Is the only page that doesn't follow the pattern /:path/index.html
+>
+> - Command line experience is required to take this course as well as shell scripting.
+> - Prior experience working on a HPC cluster. If you are new to a HPC cluster we recommend participants to go through [hpc-intro](https://hpc-carpentry.github.io/hpc-intro/).
+> - You should be familiar with working with LAMMPS and understand its input files and be able to run LAMMPS jobs in an HPC cluster.
```
They don't need to be able to run LAMMPS jobs on an HPC system...that is what we are supposed to be teaching them!

cilwerner committed:
Updated _config.yml based on review comments

exec_env abstractions for Kokkos and GPU may need adjusting based on yml formatting

cilwerner committed:
testing exec_env abstractions

cilwerner committed:
exec_env testing with indents for multi-line abstractions

cilwerner committed:
exec_env with \n

cilwerner committed:
exec_env with blank lines

cilwerner committed:
exec_env \n in whitespace

cilwerner committed:
exec_env literal style with pipes

cilwerner committed:
exec_env | and ""

cilwerner committed:
formatting test for abstractions in index

cilwerner committed:
test with additional code marks

cilwerner committed:
formatting issue

cilwerner committed:
different callout used

cilwerner committed:
testing abstraction in different file

cilwerner committed:
revert to working version of index.md

cilwerner committed:
changed to using && to solving yml multi-line issue

cilwerner committed:
test no &

cilwerner committed:
test build

cilwerner committed:
corrected format error
